proposals/0046-actions.md

```diff
@@ -0,0 +1,707 @@
+      onSubmit={async (e) => {
+        e.preventDefault();
+        const formData = new FormData(e.target);
+        const { error } = await actions.blog.comment(formData);
+        if (error && isInputError(error)) {
+          setBodyError(error.fields.body);
+        }
+        // handle result
+      }}
+    >
+      <textarea rows={10} name="body"></textarea>
+      {bodyError && <p class="error">{bodyError}</p>}
+      <button type="submit">Post</button>
+    </form>
+  );
+}
+```
+
+## Custom errors
+
+You may need to raise an exception from your action `handler()`. For this, `astro:actions` provides an `ActionError` object. This includes a `code` to set a human-readable status code like `'BAD_REQUEST'` or `'FORBIDDEN'`. These match the codes supplied by the [tRPC error object.](https://trpc.io/docs/server/error-handling) `ActionError` also includes an optional `message` property to pass information about the error.
+
+This example creates an `updateUsername` action, and raises an `ActionError` with the code `NOT_FOUND` when a given user ID is not found:
+
+```ts
+import { defineAction, ActionError } from "astro:actions";
+import { db, User, eq } from "astro:db";
+
+export const server = {
+  updateUsername: defineAction({
+    input: z.object({ id: z.string(), name }),
+    handler: async ({ id, name }) => {
+      const result = await db.update(User).set({ name }).where(eq(User.id, id));
+      if (result.rowsAffected === 0) {
+        throw new ActionError({
+          code: "NOT_FOUND",
+          message: `User with id ${id} not found.`,
+        });
+      }
+      return id;
+    },
+  }),
+};
+```
+
+To handle this error, you can call your action and check whether an `error` is present. This property will be of type `ActionError`.
+
+## Access API Context
+
+You can access the Astro API context from the second parameter to your action `handler()`. This grants access to the base `request` object, cookies, middleware `locals` and more.
+
+This example gates `getUser()` requests by checking for a session cookie. If the user is unauthorized, you can raise a "bad request" error for the client to handle.
+
+```ts
+// src/actions/index.ts
+import { defineAction, ActionError, z } from "astro:actions";
+import { db, Comment, Likes, eq, sql } from "astro:db";
+
+export const server = {
+  getUsers: defineAction({
+    handler: async (_, context) => {
+      if (!context.cookies.has('expected-session')) {
+        throw new ActionError({
+          code: "UNAUTHORIZED",
+        });
+      }
+      // return users
+    },
+  }),
+};
+```
+
+## Call actions directly from server code
+
+You may need to call an action handler directly from server pages and endpoints. To do so, use the `Astro.callAction()` function. Pass the function you want to call as the first parameter, and any input arguments as the second parameter:
+
+```astro
+---
+import { actions } from 'astro:actions';
+
+const result = await Astro.callAction(actions.searchPosts, {
```

Comment:
It would feel more natural to call `action.searchPosts()` directly on the server too, similar to how it's done on the client.

Reply:
Ah, this was removed recently due to technical limitations of injecting `APIContext`. This _did_ work in an earlier version, but it was removed due to environment issues with async local storage. I agree it's a compromise, but we unfortunately don't have alternatives today.

Reply:
Ah that's a bit unfortunate. Maybe one day!

This is a very minor issue anyway, since the big problem/use-case that actions solve is calling backend code from frontend.

Reply:
@mayank99 Yeah, that was my thinking. I agree that "just call it" is the dream! There's an avenue using compiler hacks that I dreamt up, but the maintenance burden was hard to justify. Open to any JS heroes that can find a path.